core/src/main/scala/cats/syntax/foldable.scala

```diff
@@ -331,6 +331,18 @@ final class FoldableOps0[F[_], A](private val fa: F[A]) extends AnyVal {
    */
   def maximumByOption[B: Order](f: A => B)(implicit F: Foldable[F]): Option[A] =
     F.maximumOption(fa)(Order.by(f))
+
+  def combineAllOption(implicit ev: Semigroup[A], F: Foldable[F]): Option[A] =
+    if (F.isEmpty(fa)) None else ev.combineAllOption(toIterable)
+
+  /**
+   * Convert F[A] to an Iterable[A].
+   *
+   * This method may be overridden for the sake of performance, but implementers should take care
+   * not to force a full materialization of the collection.
+   */
+  def toIterable(implicit F: Foldable[F]): Iterable[A] =
+    F.foldRight[A, Stream[A]](fa, Eval.now(Stream.empty))((a, eb) => eb.map(Stream.cons(a, _))).value
```

Comment:
@travisbrown  

Hello, sorry to bother you again on this PR.

I was looking at the `toIterable` implementation to understand why would we use Stream.

These are the possible arguments that I read from the original PR-
> Still aiming at mitigating the possible cost of building a list, it could be delayed up to when the iterator is traversing it

and 

>@travisbrown .toList.iterator forces a full materialization, and we don't control the API of List. So, the uses of a lazy structure (like Iterator, Stream, LazyList) are very different since we can interact with other tools that will not force a full materialization of the data. This is useful in things like spark or scalding.

> That said, I think in a library focused on FP, we should use an immutable lazy structure if possible, so I would lean towards Stream/LazyList. Users can then call .iterator on that, but we don't control that API. Alternatively, they can use foldLeft to tear down the Stream without materializing the whole thing in memory at once.

this is when I added a few `printLn` in between, confirming my suspicion that the Stream is created from the end to the first element, this means that the entire Stream is already materialised yet not visible.

<img width="397" alt="Screenshot 2020-03-05 at 9 09 28 PM" src="https://user-images.githubusercontent.com/8997846/75985134-ae8b9280-5f26-11ea-97a2-ac6eb871dd7b.png">

when replacing Stream with List, no difference.


<img width="262" alt="Screenshot 2020-03-05 at 9 16 09 PM" src="https://user-images.githubusercontent.com/8997846/75985137-b1868300-5f26-11ea-9833-d798b3f573c2.png">

and given this evidence we can replace this entire implementation with-

`F.foldLeft(fa, List.empty[A])((acc, a) => a :: acc).reverse`

Stream is materialised yet not visible then re-materialised (going through the whole `synchronise` thing).

Is my analysis correct?

Reply:
@travisbrown if you get a chance to go through this one